aspnet/mvc/controllers/filters.rst

```diff
@@ -1,8 +1,312 @@
+
+.. image:: filters/_static/filter-pipeline-1.png
+
+Different kinds of filters run at different points within the pipeline. Some filters, like authorization filters, only run before the next stage in the pipeline, and take no action afterward. Other filters, like action filters, can execute both before and after other parts of the pipeline execute, as shown below.
+
+.. image:: filters/_static/filter-pipeline-2.png
+
+Which filter do I need?
+^^^^^^^^^^^^^^^^^^^^^^^
+
+:ref:`Authorization filters <authorization-filters>` are used to determine whether the current user is authorized for the request being made.
+
+:ref:`Resource filters <resource-filters>` are the first filter to handle a request after authorization, and the last one to touch the request as it is leaving the filter pipeline. They're especially useful to implement caching or otherwise short-circuit the pipeline for performance reasons.
+
+:ref:`Action filters <action-filters>` wrap calls to individual action method calls, and can manipulate the arguments passed into an action as well as the response returned from it.
+
+:ref:`Exception filters <exception-filters>` are used to apply global policies to unhandled exceptions in the MVC app.
+
+:ref:`Result filters <result-filters>` wrap the execution of individual action results, but only when no exceptions have been thrown. They are ideal for logic that must surround view execution or formatter execution.
+
+Implementation
+^^^^^^^^^^^^^^
+
+All filters support both synchronous and asynchronous implementations through different interface definitions. Choose the sync or async variant depending on the kind of task you need to perform. They are interchangeable from the framework's perspective.
+
+Synchonous filters define both an On\ *Stage*\ Executing and On\ *Stage*\ Executed method (with noted exceptions). The On\ *Stage*\ Executing method will be called before the event pipeline stage by the Stage name, and the On\ *Stage*\ Executed method will be called after the pipeline stage named by the Stage name.
+
+.. literalinclude:: filters/sample/src/FiltersSample/Filters/SampleActionFilter.cs
+  :language: c#
+  :emphasize-lines: 6,8,13
+
+Asynchronous filters define a single On\ *Stage*\ ExecutionAsync method that will surround execution of the pipeline stage named by Stage. The On\ *Stage*\ ExecutionAsync method is provided a *Stage*\ ExecutionDelegate delegate which will execute the pipeline stage named by Stage when invoked and awaited.
+
+.. literalinclude:: filters/sample/src/FiltersSample/Filters/SampleAsyncActionFilter.cs
+  :language: c#
+  :emphasize-lines: 6,8-9
+
+In most cases, you can implement both the async and synchronous interfaces with the same class, like so:
```

Comment:
In most cases, you can implement both the async and synchronous interfaces with the same class:
[you don't need like so - if that feels incomplete, use something similar to 
as shown below: